### PR TITLE
Add support for COPY --exclude and ADD --exclude options

### PR DESCRIPTION
**Pull request by rhatdan**
Fixes: https://github.com/containers/buildah/issues/5678

#### What type of PR is this?

> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test 
> /kind feature
> /kind flake
> /kind other

#### Does this PR introduce a user-facing change?

```release-note
COPY option  and buildah copy now support --exclude options.
```

### CODE REVIEW

File: cmd/buildah/addcopy.go [hunk 1]
```diff
@@ -20,23 +20,24 @@ import (
...
 	removeSignatures bool
-	signaturePolicy  string
-	authfile         string
-	creds            string
-	tlsVerify        bool
-	certDir          string
 	retry            int
 	retryDelay       string
+	signaturePolicy  string
+	tlsVerify        bool
```
Comment (nalind) [inline cmd/buildah/addcopy.go:40]: This makes it it harder to find the specific commits where fields were added.  Please don't do it.

File: docs/buildah-add.1.md [hunk 2]
```diff
@@ -49,6 +49,11 @@ Build context directory. Specifying a context directory causes Buildah to
 chroot into that context directory. This means copying files pointed at
 by symbolic links outside of the chroot will fail.
 
+**--exclude** *pattern*
+
+Exclude coping files mattching the specified pattern. Option can be specified
```
Comment (nalind) [inline docs/buildah-add.1.md:54]: Typos: `coping`, `mattching`.

File: docs/buildah-copy.1.md [hunk 3]
```diff
@@ -48,6 +48,11 @@ Build context directory. Specifying a context directory causes Buildah to
 chroot into the context directory. This means copying files pointed at
 by symbolic links outside of the chroot will fail.
 
+**--exclude** *pattern*
+
+Exclude coping files mattching the specified pattern. Option can be specified
```
Comment (nalind) [inline docs/buildah-copy.1.md:53]: Typos: `coping`, `mattching`.

**Comment by rhatdan** [comment]
@flouthoc @giuseppe @nalind @mheon PTAL

**Comment by rhatdan** [comment]
@slinkydeveloper PTAL

**Comment by rhatdan** [comment]
@nalind @flouthoc @giuseppe @TomSweeneyRedHat @lsm5 @mheon PTAL

**Review by flouthoc — APPROVED** [review]
LGTM

**Comment by nalind** [comment]
This needs quite a few more tests to ensure that we're getting this right.  Add conformance tests using multiple ways of expressing the sources to be copied (".", "/", "*"), and tests that combine `COPY` or `ADD` using `--exclude` with an ignorefile, including cases where they disagree on whether or not a particular item should or should not be excluded.

**Comment by nalind** [comment]
Set `dockerUseBuildKit` in the test case to have the client set the right flag in its request to the REST API's build endpoint.  I'm not 100% sure that excludes are handled correctly between the client and the daemon and between the daemon and buildkit, but then maybe the odd things I'm seeing are somehow specific to my system.

**Comment by TomSweeneyRedHat** [comment]
LGTM once comments are addressed and tests are hip

**Comment by villesau** [comment]
Is there timeline for landing this? 👀

**Comment by rhatdan** [comment]
@villesau I have been working on other tools, have not come back to change the code that @nalind suggests.

If you want to try to change it go for it.

The actual code is ready to go, the tests are just a little broken.

**Comment by villesau** [comment]
@rhatdan Are there some instructions how to set this up and run the tests? I might try, but please do not count on it 😅

File: tests/conformance/conformance_test.go [hunk 4]
```diff
@@ -2444,6 +2444,83 @@ var internalTestCases = []testCase{
...
+		compatScratchConfig: types.OptionalBoolFalse,
+		dockerUseBuildKit:   true,
+	},
+
+	{
+		name: "copy--exclude-includes-dot-slash",
+		dockerfileContents: strings.Join([]string{
+			"# syntax=docker/dockerfile:1.9-labs",
+			"FROM scratch",
+			"COPY --exclude='!**/*-c' ./ subdir/",
```
Comment (nalind) [inline tests/conformance/conformance_test.go:2503]: (Turns out I was mistaken, and quoting does work for the value of the `--exclude=` argument.)

File: docs/buildah-add.1.md [hunk 5]
```diff
@@ -49,6 +49,11 @@ Build context directory. Specifying a context directory causes Buildah to
 chroot into that context directory. This means copying files pointed at
 by symbolic links outside of the chroot will fail.
 
+**--exclude** *pattern*
+
+Exclude copying files matching the specified pattern. Option can be specified
+multiple times.
```
Comment (nalind) [inline docs/buildah-add.1.md:55]: Perhaps a reference to the format of patterns being described in `containerignore(5)` from the `common` package?

File: docs/buildah-copy.1.md [hunk 6]
```diff
@@ -48,6 +48,11 @@ Build context directory. Specifying a context directory causes Buildah to
 chroot into the context directory. This means copying files pointed at
 by symbolic links outside of the chroot will fail.
 
+**--exclude** *pattern*
+
+Exclude copying files matching the specified pattern. Option can be specified
+multiple times.
```
Comment (nalind) [inline docs/buildah-copy.1.md:54]: Perhaps a reference to the format of patterns being described in `containerignore(5)` from the `common` package?

File: imagebuildah/stage_executor.go [hunk 7]
```diff
@@ -373,10 +373,7 @@ func (s *StageExecutor) Copy(excludes []string, copies ...imagebuilder.Copy) err
 			return errors.New("COPY --parents is not supported")
 		}
 		if len(cp.Excludes) > 0 {
-			if cp.Download {
-				return errors.New("ADD --excludes is not supported")
-			}
-			return errors.New("COPY --excludes is not supported")
+			excludes = append(excludes, cp.Excludes...)
```
Comment (nalind) [inline imagebuildah/stage_executor.go:376]: Append to `slices.Clone(excludes)`, please.  We don't own `excludes` and don't want to inadvertently modify the slice that was passed to us.

File: imagebuildah/stage_executor.go [hunk 8]
```diff
@@ -1325,12 +1322,12 @@ func (s *StageExecutor) Execute(ctx context.Context, base string) (imgID string,
 		// Also check the chmod and the chown flags for validity.
 		for _, flag := range step.Flags {
 			command := strings.ToUpper(step.Command)
-			// chmod, chown and from flags should have an '=' sign, '--chmod=', '--chown=' or '--from='
-			if command == "COPY" && (flag == "--chmod" || flag == "--chown" || flag == "--from") {
-				return "", nil, false, fmt.Errorf("COPY only supports the --chmod=<permissions> --chown=<uid:gid> and the --from=<image|stage> flags")
+			// chmod, chown and from flags should have an '=' sign, '--chmod=', '--chown=' or '--from=' or '--exclude='
+			if command == "COPY" && (flag == "--chmod" || flag == "--chown" || flag == "--from" || flag == "--exclude") {
+				return "", nil, false, fmt.Errorf("COPY only supports the --chmod=<permissions> --chown=<uid:gid> and the --from=<image|stage> --exclude=<pattern> flags")
```
Comment (nalind) [inline imagebuildah/stage_executor.go:1327]: The "and the" should be later in the error message

**Review by giuseppe — APPROVED** [review]
LGTM

**Pull request merged**